cognite_toolkit/cdf_tk/templates/data_classes/_project_directory.py

```diff
@@ -0,0 +1,298 @@
+                f"  No backup {'would have been' if self._dry_run else 'will be'} done."
+            )
+            return
+        elif self._has_changed_cognite_modules:
+            print("[bold yellow]WARNING:[/] The cognite_modules have changed, it will not be upgraded.")
+
+        if not no_backup:
+            prefix = "Would have backed up" if self._dry_run else "Backing up"
+            if verbose:
+                print(f"{prefix} {self.target_dir_display}")
+            if not self._dry_run:
+                backup_dir = tempfile.mkdtemp(prefix=f"{self.project_dir.name}.", suffix=".bck", dir=Path.cwd())
+                shutil.copytree(self.project_dir, Path(backup_dir), dirs_exist_ok=True)
+        else:
+            print(
+                "[bold yellow]WARNING:[/] --no-backup is specified, no backup "
+                f"{'would have been' if self._dry_run else 'will be'} done."
+            )
+
+    def print_what_to_copy(self) -> None:
+        if not self._has_changed_cognite_modules:
+            print("  Will upgrade modules and files in place.")
+            super().print_what_to_copy()
+
+    def copy(self, verbose: bool) -> None:
+        if self._has_changed_cognite_modules:
+            return
+        else:
+            super().copy(verbose)
+
+    def set_source(self, git_branch: str | None) -> None:
+        if git_branch is None:
+            return
+
+        self._source = self._download_templates(git_branch, self._dry_run)
+
+    def upsert_config_yamls(self, clean: bool) -> None:
+        if clean:
+            super().upsert_config_yamls(clean)
+            return
+        if self._has_changed_cognite_modules:
+            return
+
+        existing_environments = list(self.project_dir.glob("config.*.yaml"))
+        if len(existing_environments) == 0:
+            print("  [bold yellow]WARNING:[/] No existing config.[env].yaml files found, creating from the defaults.")
+            super().upsert_config_yamls(clean)
+            return
+
+        config_yamls = ConfigYAMLs.load_existing_environments(existing_environments)
+
+        config_yamls.load_default_variables(self._source)
+        config_yamls.load_variables(self._source)
+
+        for environment, config_yaml in config_yamls.items():
+            config_filepath = self.project_dir / f"config.{environment}.yaml"
+            for entry in config_yaml.added:
+                print(entry)
+            for entry in config_yaml.removed:
+                print(entry)
+
+            if self._dry_run:
+                print(f"Would write {config_filepath.name!r} to {self.target_dir_display}")
+            else:
+                config_filepath.write_text(config_yaml.dump_yaml_with_comments(indent_size=2))
+                print(f"Wrote {config_filepath.name!r} file to {self.target_dir_display}")
+
+    def done_message(self) -> str:
+        return f"Automatic upgraded of {self.target_dir_display} is done.\nManual Steps:"
```

Comment:
There's a lot of text and this disappears. You need to make it much more prominent (with bold and green maybe?).
Also, it is not clear whether the manual steps need to be done in addition or if they are just a description of what has been done (and that you should do the manual steps to your own custom modules).
I suggest you create a frame with this info, maybe even make it default interactive, so you can step through first the auto-upgrade, then individual changes?

Reply:
Tried to improve it now.